keps/sig-cluster-lifecycle/wgs/20191001-component-config-validation-interface.md

```diff
@@ -0,0 +1,177 @@
+Currently, some cluster lifecycle tools vendor in and use Go code from the target components themselves to be able to verify component configs early enough during the provisioning operation, so that meaningful error messages are displayed in time directly to the user. This, however, poses its own problems as such pieces of code can deviate from the actual shipped component and can lead to false positives and false negatives.  
+Hence, it’s best that validation is performed by the component itself in a one off operation.
+
+
+### Goals
+
+The main goal is to provide a stable standard command line interface for components that can be used to validate component configs and retrieve appropriate errors and warnings.
+
+
+### Non-Goals
+
+To impose the implementation of this proposal on all Kubernetes components, that provide component config. Implementing this interface is entirely optional.
+
+
+## Proposal
+
+The proposal is to introduce a new standardized sub-command for each component that implements the interface. This command can then be called by invoking the component’s executable or creating a container with it.
+
+
+### The “experimental-config-validate“ sub-command
+
+The new sub-command will be known as “experimental-config-validate”. Upon graduation to beta and GA, this sub-command will be renamed to just “config-validate”.
+
+It must support a couple of mandatory flags:
+- “--help” or “-h” - if specified, will dump usage information of the sub-command and exit without doing any config validation.
+- “--file” or “-f” - supplies a YAML or JSON file containing a component config to be validated.
+
+In all implementations of “experimental-config-validate”, the “--file” flag can be specified multiple times. If a component does not support multi-file config, supplying multiple “--file” values must return an appropriate error to the user without performing any validation.  
+If no “--file” or “--help” switches are supplied, the command reads the config to be validated from the standard input.
+
+When the “experimental-config-validate” is executed supplied with a config, it should validate it and:
+- Exit with success if the supplied config is found to be valid. Each line of text printed on the standard output is considered to be a single warning message. Multiple warning messages can be printed.
+- Exit with failure if an error is encountered and the component will be unable to operate normally with the config as is. Each line of text printed on the standard output is considered to be a single error message. Multiple error messages can be printed.
+
+
+### Sub-command placement
+
+The “experimental-config-validate” sub-command must be placed under the command that houses the component whose config needs to be validated.  
+Below is an example that uses kubelet:
+```
+# kubelet experimental-config-validate --file=config.yaml
```

Comment:
We don't have sub-commands at all today. Does the new subcommand work with sub-command-less execution?

Reply:
Alternative: `--validate-only`-like flag.

Reply:
My choice of a sub-command was to allow for better encapsulation of this interface. A flag is going to introduce more confusion in already different commands in the different components.
On the other hand, sub-commands are not ideal either. Given the fact, that we have a component called `foo`, it's command line syntax can be something like this:
```
foo [flags] <command-line-args>
```

Where `<command-line-args>` is a set of non-flag arguments (host names, paths to required files, etc.). The introduction of a sub-command is going to disable the use of the sub-command name as an argument (you won't be able to have a host named `experimental-config-validate`).
Worse still, if someone mistypes the name of the sub-command, it will end as an argument of the base command.

There is also a third option. That is to not specify a common interface at all, but rather say "any interface that should be able to do this and that on the command line". This is probably going to be best received by implementers, but will likely generate different CLIs and diversion between components.

I am glad, that we are starting a debate here.

Reply:
I am not against a subcommand. My question is whether cobra can have subcommands and a non-subcommand mode at the same time.

Reply:
> A flag is going to introduce more confusion in already different commands in the different components.

Which components actually expose other subcommands (other than kubeadm)? Kubelet has none, for example, so --validate-config might be more reasonable.

There's also the question of whether we should do --validate-config=file or just do --validate-config and rely on --config _and_ flags to specify the options (e.g. should we also take flags into account). The latter is basically a --dry-run option.